web/src/types.ts

```diff
@@ -6,19 +6,111 @@ export type MatchedMessage = {
 
 // TODO/typescript: Move this to message_store
 export type MessageType = "private" | "stream";
+export type MessageContentType = "text/html" | "text/x-markdown";
+export type MessageReactionType = "unicode_emoji" | "realm_emoji" | "zulip_extra_emoji";
+
+// TODO/typescript: Move this to server_events
+export type TopicLink = {
+    text: string;
+    url: string;
+};
+
+// TODO/typescript: Move these types to message_store
+
+export type DisplayRecipientUser = {
+    email: string;
+    full_name: string;
+    id: number;
+    is_mirror_dummy: boolean;
+    unknown_local_echo_user?: boolean;
+};
+
+export type DisplayRecipient = string | DisplayRecipientUser[];
+
+export type MessageEditHistoryEntry = {
+    user_id: number | null;
+    timestamp: number;
+    prev_content?: string;
+    prev_rendered_content?: string;
+    prev_rendered_content_version?: number;
+    prev_stream?: number;
+    prev_topic?: string;
+    stream?: number;
+    topic?: string;
+};
+
+export type MessageReaction = {
+    emoji_name: string;
+    emoji_code: string;
+    reaction_type: MessageReactionType;
+    user_id: number;
+    user: DisplayRecipientUser;
+};
 
 // TODO/typescript: Move this to message_store
 export type RawMessage = {
+    avatar_url: string;
+    client: string;
+    content: MessageContentType;
+    content_type: string;
+    display_recipient: DisplayRecipient;
+    edit_history?: MessageEditHistoryEntry[];
+    id: number;
+    is_me_message: boolean;
+    last_edit_timestamp?: number;
+    reactions: MessageReaction[];
+    recipient_id: number;
     sender_email: string;
-    stream_id: number;
+    sender_full_name: string;
+    sender_id: number;
+    sender_realm_str: string;
+    stream_id?: number;
     subject: string;
+    submessages: string[];
+    timestamp: number;
+    topic_links: TopicLink[];
     type: MessageType;
+    flags: string[];
 } & MatchedMessage;
 
 // TODO/typescript: Move this to message_store
-export type Message = RawMessage & {
-    to_user_ids: string;
+export type MessageCleanReaction = {
+    class: string;
+    count: number;
+    emoji_alt_code: string;
+    emoji_code: string;
+    emoji_name: string;
+    is_realm_emoji: boolean;
+    label: string;
+    local_id: string;
+    reaction_type: string;
+    user_ids: number[];
+    vote_text: string;
+};
+
+// TODO/typescript: Move this to message_store
+export type Message = Omit<RawMessage, "reactions"> & {
```

Comment:
I propose that we can group the following properties by where they are added, not necessarily into different types, at least for now. For example, we can group the booleans added by `set_message_booleans` with a single line comment above mentioning the source.